### ISSUE TITLE
Feeding a value for placeholder tensor in Tensorflow C++ API

### ISSUE THREAD

**Issue by umairalipathan** [tensorflow/serving#518]
I have retrained Inception-v3 model using (Tensorflow) Python API and saved a standalone Graph in .pb by modifying tensorflow/tensorflow/examples/image_retraining/retrain.py with the following modification to add a drop out layer before the classification layer:
`def nn_layer(input_tensor, input_dim, output_dim, layer_name, activation_name='activation', act=tf.nn.softmax):`
    `# Adding a name scope ensures logical grouping of the layers in the graph.`
    `with tf.name_scope(layer_name):`
          ` # This Variable will hold the state of the weights for the layer`
      ` with tf.name_scope('weights'):`
          `weights = weight_variable([input_dim, output_dim])`
          `variable_summaries(weights, layer_name + '/weights')`
      `with tf.name_scope('dropout'):`
          `keep_prob = tf.placeholder(tf.float32)`
          `tf.scalar_summary('dropout_keep_probability', keep_prob)`
          `drop = tf.nn.dropout(input_tensor, keep_prob)`
          ` variable_summaries(drop, layer_name + '/dropout')`
     ` with tf.name_scope('biases'):`
          `biases = bias_variable([output_dim])`
          `variable_summaries(biases, layer_name + '/biases')`
      `with tf.name_scope('Wx_plus_b'):`
         ` preactivate = tf.matmul(drop, weights) + biases`
          `tf.histogram_summary(layer_name + '/pre_activations', preactivate)`
         `# activations = act(preactivate, 'activation')`
      `with tf.name_scope(activation_name):`
          `activations = act(preactivate)`
          `tf.histogram_summary(layer_name + '/activations', activations)`
    `return preactivate, activations, keep_prob`

The code to generate predictions in python is as follows:

`softmax_tensor = sess.graph.get_tensor_by_name('final_layer/final_result/Softmax:0')`
`predictions = sess.run(softmax_tensor, { 'DecodeJpeg/contents:0': image_data, 'final_layer/dropout/Placeholder:0': 1.})`

The C++ counterpart of the python code is as follows:

`string input_layer = "Mul"; `
`string output_layer = "final_layer/dropout/Placeholder:0";`
`Status run_status = session->Run({{input_layer, resized_tensor}}, {output_layer}, {}, &outputs);`

The C++ code ends up with the following error message:

`Running model failed: Invalid argument: You must feed a value for placeholder tensor 'final_layer/dropout/Placeholder'`

What should I change in the above C++ code to remove this error? In other words, how do I change a placeholder value in the C++ code as in python code. I have already asked this question on stackoverflow, just to no avail. Any help will be highly appreciated.

**Comment by chrisolston** [comment]
(Meta: IIUC you are using tensorflow directly (Session::Run()) and not tensorflow-serving. That's fine but it's worth pointing that out so that other readers of this thread don't get confused, since it's posted in the tf-serving forum. tf-serving would add a level of indirection on top of the Session::Run() API.)

It should just be a matter of adding a tensor to the "inputs" vector of your Session::Run() invocation, as in:
Status run_status = session->Run({{input_layer, resized_tensor}, {placeholder_tensor_name, placeholder_tensor_data}}, {output_layer}, {}, &outputs);

**Comment by umairalipathan** [comment]
Thanks @chrisolston. Considering the python counterpart of the C++ API, could you please tell me how do I define placeholder_tensor_name and placeholder_tensor_data?

**Comment by chrisolston** [comment]
placeholder_tensor_name is just the string name of the tensor: "final_layer/dropout/Placeholder" I think.
placeholder_tensor_data is the tensor: "keep_prob"

**Comment by umairalipathan** [comment]
Yeah, I am trying the same. But it still gives the same error:
`Running model failed: Invalid argument: You must feed a value for placeholder tensor 'final_layer/dropout/Placeholder' with dtype float`
` [[Node: final_layer/dropout/Placeholder = Placeholder[dtype=DT_FLOAT, shape=[], _device="/job:localhost/replica:0/task:0/gpu:0"]()]]`
` [[Node: final_layer/dropout/Placeholder/_1 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_4_final_layer/dropout/Placeholder", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]`

**Comment by umairalipathan** [comment]
This is the code I am using: 
`string input_layer = "final_layer/final_result/Softmax:0";`
`string output_layer = "final_layer/dropout/Placeholder:0";`
`tensorflow::Tensor keep_prob(tensorflow::DT_FLOAT, tensorflow::TensorShape());`
`keep_prob.scalar<float>()() = 1.0;`
`Status run_status = session->Run({{input_layer, input_tensor},{"final_layer/dropout/Placeholder",keep_prob}}, {output_layer}, {}, &outputs);`

**Comment by chrisolston** [comment]
You could try adding ":0" to the tensor name. Beyond that I'm really not sure. Try asking in the core tensorflow forum?

**Comment by umairalipathan** [comment]
This solves the problem. Thanks a lot.
